xfer/contrib/xfer_leap/leap_meta_repurposer.py

```diff
@@ -0,0 +1,160 @@
+# ==============================================================================
+import random
+
+from mxnet import nd, autograd, gluon
+import mxnet as mx
+
+from .meta_repurposer import MetaRepurposer
+from .leap import Leap
+from .metalogger import MetaLogger
+
+DEFAULT_META_OPTIMIZER_PARAMS = {'learning_rate': 0.0001}
+DEFAULT_TASK_OPTIMIZER_PARAMS = {'learning_rate': 0.005}
+
+
+class LeapMetaRepurposer(MetaRepurposer):
+    """
+    Repurposer to do meta-learning using Leap algorithm
+
+    :param input_model: Source neural network to use for meta-learning.
+    :type input_model: :class:`mxnet.gluon.Block`
+    """
+
+    def __init__(self, model, num_meta_steps, num_epochs, loss_class=gluon.loss.L2Loss,
+                 meta_optimizer='sgd', meta_optimizer_params=None, task_optimizer='sgd', task_optimizer_params=None,
+                 log_params=False, verbosity=1, context=None):
+        self.model = model
+        self.meta_logger = MetaLogger()
+
+        self.num_meta_steps = num_meta_steps
+        self.num_epochs = num_epochs
+
+        self.context = context
+        if self.context is None:
+            self.context = mx.cpu()
+
+        self.loss_class = loss_class
+        self.meta_optimizer = meta_optimizer
+        self.meta_optimizer_params = meta_optimizer_params
+        if self.meta_optimizer_params is None:
+            self.meta_optimizer_params = DEFAULT_META_OPTIMIZER_PARAMS
+        self.task_optimizer = task_optimizer
+        self.task_optimizer_params = task_optimizer_params
+        if self.task_optimizer_params is None:
+            self.task_optimizer_params = DEFAULT_TASK_OPTIMIZER_PARAMS
+
+        self.log_params = log_params
+        self.verbosity = verbosity
+
+    @staticmethod
+    def _task_sampler(train_tasks, meta_batch_size, with_replacement=False):
+        """
+        Tasks' meta-batch sampler
+
+        :param train_tasks: List of enumerated training tasks
+        :param meta_batch_size: Number of tasks per meta-batch
+        :param with_replacement: If True, sample tasks with replacement.
+        :return: List of sampled enumerated tasks
+        """
+
+        n_tasks = len(train_tasks)
+
+        if n_tasks == 1:
+            tasks = train_tasks * meta_batch_size
+        elif with_replacement:
+            tasks = [train_tasks[random.randint(0, n_tasks - 1)] for _ in range(meta_batch_size)]
+        else:
+            tasks = []
+            ix_tasks = list(range(n_tasks))
+            while True:
+                random.shuffle(ix_tasks)
+                tasks.extend([train_tasks[i] for i in ix_tasks])
+                if len(tasks) >= meta_batch_size:
+                    break
+            tasks = tasks[:meta_batch_size]
+
+        return tasks
+
+    def repurpose(self, train_tasks, meta_batch_size=None, with_replacement=False):
+        """
+        Meta-learn parameters of the input model such that they are optimized for the provided train_tasks.
```

Comment:
Starting from the parameters of input_model as the initial point, the Leap algorithm is applied. It learns a final parametrization of input_model that can be quickly adapted to solve a test task sampled from the same underlying distribution that generated train_tasks.